fdbclient/RESTSimKmsVault.actor.cpp

```diff
@@ -0,0 +1,510 @@
+}
+
+Optional<int64_t> getExpireInterval(Optional<int64_t> refTS, const int64_t defaultTtl) {
+	ASSERT(refTS.present());
+
+	if (BUGGIFY) {
+		return Optional<int64_t>(-1);
+	}
+	return (refTS.get() + defaultTtl);
+}
+
+void addErrorToDoc(rapidjson::Document& doc, std::vector<ErrorDetail>& details) {
+	// Add error details
+	rapidjson::Value errorDetails(rapidjson::kArrayType);
+
+	for (const auto& detail : details) {
+		rapidjson::Value errorDetail(rapidjson::kObjectType);
+
+		if (!detail.errorMsg.empty()) {
+			// Add "errorMsg"
+			rapidjson::Value key(ERROR_MSG_TAG, doc.GetAllocator());
+			rapidjson::Value errMsg;
+			errMsg.SetString(detail.errorMsg.data(), detail.errorMsg.size(), doc.GetAllocator());
+			errorDetail.AddMember(key, errMsg, doc.GetAllocator());
+		}
+		if (!detail.errorCode.empty()) {
+			// Add "value" - token value
+			rapidjson::Value key(ERROR_CODE_TAG, doc.GetAllocator());
+			rapidjson::Value errCode;
+			errCode.SetString(detail.errorCode.data(), detail.errorCode.size(), doc.GetAllocator());
+			errorDetail.AddMember(key, errCode, doc.GetAllocator());
+		}
+		errorDetails.PushBack(errorDetail, doc.GetAllocator());
+	}
+
+	// Append "error"
+	rapidjson::Value key(ERROR_TAG, doc.GetAllocator());
+	doc.AddMember(key, errorDetails, doc.GetAllocator());
+}
+
+void prepareErrorResponse(Reference<HTTP::IncomingRequest> request,
+                          Reference<HTTP::OutgoingResponse> response,
+                          const std::string& errMsg,
+                          const std::string& errCode,
+                          Optional<int> version = Optional<int>()) {
+	response->code = HTTP::HTTP_STATUS_CODE_OK;
+	response->data.headers = request->data.headers;
+
+	rapidjson::Document doc;
+	doc.SetObject();
+
+	if (version.present()) {
+		addVersionToDoc(doc, version.get());
+	}
+
+	std::vector<ErrorDetail> errors;
+	errors.emplace_back(errCode, errMsg);
+	addErrorToDoc(doc, errors);
+
+	// Serialize json to string
+	rapidjson::StringBuffer sb;
+	rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+	doc.Accept(writer);
+
+	std::string str(sb.GetString(), sb.GetSize());
+	response->data.contentLen = str.size();
+	PacketWriter pw(response->data.content->getWriteBuffer(str.size()), nullptr, Unversioned());
+	pw.serializeBytes(str);
+}
+
+int extractVersion(Reference<HTTP::IncomingRequest> request,
+                   Reference<HTTP::OutgoingResponse> response,
+                   const rapidjson::Document& doc) {
+	// check version tag sanityrest_malformed_response
+	if (!doc.HasMember(REQUEST_VERSION_TAG) || !doc[REQUEST_VERSION_TAG].IsInt()) {
+		prepareErrorResponse(request, response, "Missing Version", "1234");
+		CODE_PROBE(true, "RESTSimKmsVault missing version");
+		throw rest_malformed_response();
+	}
+
+	const int version = doc[REQUEST_VERSION_TAG].GetInt();
+	if (version < 0 || version > 1) {
+		prepareErrorResponse(request, response, "Invalid version", "1234");
```

Comment:
What's the purpose of preparing an http response if we then immediately throw an error?
The http response only gets sent over the network back if the request handler function returns without error.
checkValidationTokens also does this

Reply:
Malformed REST responses are _successful_ from REST communication perspective, RESTKmsConnector will get a `200` HTTP response and would parse the error details the print for records, throw appropriate error etc. 
Throwing error allows the server to break the execution context and respond with the error response.

Reply:
Ok so basically the contract is that the request handler will not send the response if the request handler function returns an error, it assumes that is an internal error and sends error code 500 back.
If you want the request handler to respond with the error message you constructed, you must just return (without error) from the request handler

Reply:
I see. Makes sense, will make necessary changes. Thanks